File: leetcode.com/python/124_Binary_Tree_Maximum_Path_Sum.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        _, maxSum = self.maxPathSumHelper(root)
        return maxSum

    def maxPathSumHelper(self, root):
        if not root:
            return (float('-inf'), float('-inf'))
        leftMaxSumAsBranch, leftMaxPathSum = self.maxPathSumHelper(root.left)
        rightMaxSumAsBranch, rightMaxPathSum = self.maxPathSumHelper(root.right)
        maxChildSumAsBranch = max(leftMaxSumAsBranch, rightMaxSumAsBranch)
        value = root.val
        maxSumAsBranch = max(maxChildSumAsBranch + value, value)
        maxSumAsRootNodeOrTriangle = max(leftMaxSumAsBranch + value + rightMaxSumAsBranch, maxSumAsBranch)
        maxPathSum = max(leftMaxPathSum, rightMaxPathSum, maxSumAsRootNodeOrTriangle)
        return (maxSumAsBranch, maxPathSum)



sol = Solution()
root = TreeNode(-3)
out = sol.maxPathSum(root)
print("Res", out)

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/124_Binary_Tree_Maximum_Path_Sum.swift
import Foundation
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func maxPathSum(_ root: TreeNode?) -> Int {
        var maxSumResult: Int = Int.min
        maxPathSumDFSHelper(root, &maxSumResult)
        return maxSumResult
    }
    
    private func maxPathSumDFSHelper(_ root: TreeNode?, _ maxSumResult: inout Int) -> Int {
        guard let root = root else {
            return 0
        }
        let leftMaxSum = maxPathSumDFSHelper(root.left, &maxSumResult)
        let rightMaxSum = maxPathSumDFSHelper(root.right, &maxSumResult)
        
        maxSumResult = [maxSumResult, root.val, root.val + leftMaxSum, root.val + rightMaxSum, root.val + leftMaxSum + rightMaxSum].max()!
        
        return [root.val, root.val + leftMaxSum, root.val + rightMaxSum].max()!
    }
}
